nevergrad/optimization/differentialevolution.py

```diff
@@ -158,6 +159,8 @@ def _internal_ask_candidate(self) -> p.Parameter:
             self.population[candidate.uid] = candidate
             self._uid_queue.asked.add(candidate.uid)
             return candidate
+        # stop queue wrapping around to lineage waiting for a tell
+        assert self._uid_queue.told, "More untold asks than population size (exceeds num_workers)"
```

Comment:
@jrapin  you are the expert for self._uid_queue.told (among so many things...), do you validate this assert ?

Reply:
Oh I guess the error is in class Portfolio. Let me propose a fix (fingers crossed :-) ).

Reply:
> @jrapin you are the expert for self._uid_queue.told (among so many things...), do you validate this assert ?

If it helps, my thinking was that there should be a tell preceding every ask after the initalization phase keeping the told queue non-empty. Even in the worst case where popsize ==num_workers and all workers are evaluating untold points, the worker that beats the others to the tell can use the same point again on the next ask.

Reply:
I've just sent a message to Jeremy, who knows that code better than anyone else and who might not have been close to github recently. Sorry for the delay; your PR is interesting.

Reply:
I used to be strict with the fact that we should not go beyond `num_workers`, but I changed my mind a couple of years ago because there are many cases you don't master all the details of what is happening (eg: a process dies and you'll never get the result), most times the user won't deal with it and we should be robust to it to simplify use. The code was then  **supposed** to be robust but visibly there are corner cases :s 
I would be therefore rather make it robust to this case (would that just take removing duplicates in UuidQueue.told ? it should be light speed so not a problem)

cc @bottler you seemed to disagree and want the user to strictly conform to the "contract", maybe we can discuss and adapt depending if I change your mind or not ;)

@Game4Move78 as a power user, would you rather it bugged explicitely, or be robust to those corner cases? (why did you happen to ask for more points?)

Reply:
I followed the hyper-parameter settings of papers that used DE for HPO and set popsize to 20 explicitly without providing num_workers, and thought it would be robust. I then asked for more points and handed them to my own adaptive resource
allocation + early stopping implementation that evaluated HPO choices multiple budgets and only provided a tell to the NG optimiser when points were either stopped early or allocated maximum budget.

This would work fine for hundreds of points until it hit that corner case with a point in the told queue that has been deleted from population. My current workaround is to provide feedback immediately on the minimum budget and then treats all evaluations on higher budgets as untold points, which works fine for DE.

If you want less strict, how about we allow duplicates in tell but at [L162](https://github.com/facebookresearch/nevergrad/blob/main/nevergrad/optimization/differentialevolution.py#L162) we add
```
while lineage not in self.population:
      lineage = self._uid_queue.ask()
```

Which I believe would toss away those deleted points.

Reply:
My personal preference to help users to master those details is to [copy Ax's client interface](https://github.com/facebook/Ax/blob/3a8f8dfd8061a78f71ec7ba697c268b4a5196520/ax/service/ax_client.py#L577) with an `abandon_tell`. For most optimisers this would just provide the optimiser with a large value, and the BO optimisers might do something different to avoid damaging the model.